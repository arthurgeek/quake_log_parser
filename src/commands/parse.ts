import { readFile } from 'node:fs/promises'
import { command } from 'cleye'
import { parse } from '@/parser/parser'

export const parseCommand = command(
  {
    name: 'parse',
    parameters: ['[file]'],
    help: {
      description: 'parse [file] and output a JSON AST',
    },
  },
  async (argv) => {
    const file = argv._.file || './data/quake.log'
    const data = await readFile(file, 'utf8')

    console.log(JSON.stringify(parse(data)))
  },
)
