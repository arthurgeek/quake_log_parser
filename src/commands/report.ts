import { readFile } from 'node:fs/promises'
import { command } from 'cleye'
import { Reporter } from '@/reporter'
import { Parser } from '@/parser/parser'

export const reportCommand = command(
  {
    name: 'report',
    parameters: ['[file]'],
    help: {
      description: 'report [file] and output JSON',
    },
  },
  async (argv) => {
    const file = argv._.file || './data/quake.log'
    const data = await readFile(file, 'utf8')
    const ast = new Parser(data).parse()

    if ('kind' in ast && ast.kind == 'AllGames') {
      console.log(JSON.stringify(new Reporter(ast).report()))
    }
  },
)
