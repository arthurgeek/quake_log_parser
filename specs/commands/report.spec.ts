import { reportCommand } from '@/commands/report'

// eslint-disable-next-line @typescript-eslint/no-empty-function
const noop = () => {}

vitest.spyOn(console, 'log')

describe('report command', () => {
  it('prints a JSON structure for all games in `data/quake.log if no file is given`', async () => {
    vitest.mocked(console.log).mockImplementation(noop)

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    await reportCommand.callback!({ _: { file: undefined } })

    expect(vitest.mocked(console.log).mock.calls).toMatchSnapshot()
  })

  it('accepts another file`', async () => {
    vitest.mocked(console.log).mockImplementation(noop)

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    await reportCommand.callback!({
      _: { file: './specs/data/smallerWithKills.log' },
    })

    expect(vitest.mocked(console.log).mock.calls).toMatchInlineSnapshot(`
      [
        [
          "[{\\"id\\":1,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Zeh\\",\\"kills\\":1},{\\"id\\":\\"3\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":2},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":0},{\\"id\\":\\"5\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":1}],\\"totalKills\\":14,\\"killsByMeans\\":{\\"MOD_ROCKET\\":4,\\"MOD_ROCKET_SPLASH\\":4,\\"MOD_TRIGGER_HURT\\":5,\\"MOD_RAILGUN\\":1}}]",
        ],
      ]
    `)
  })
})
