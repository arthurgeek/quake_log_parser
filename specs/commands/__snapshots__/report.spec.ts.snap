// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`report command > prints a JSON structure for all games in \`data/quake.log if no file is given\` 1`] = `
[
  [
    "[{\\"id\\":1,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":0}],\\"totalKills\\":0},{\\"id\\":2,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":-5},{\\"id\\":\\"3\\",\\"name\\":\\"Mocinha\\",\\"kills\\":0}],\\"totalKills\\":11},{\\"id\\":3,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":-1},{\\"id\\":\\"3\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":1},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":-2}],\\"totalKills\\":4},{\\"id\\":4,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":13},{\\"id\\":\\"3\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":19},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":20},{\\"id\\":\\"5\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":13}],\\"totalKills\\":105},{\\"id\\":5,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Zeh\\",\\"kills\\":1},{\\"id\\":\\"3\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":2},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":0},{\\"id\\":\\"5\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":1}],\\"totalKills\\":14},{\\"id\\":6,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":8},{\\"id\\":\\"3\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":3},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":7},{\\"id\\":\\"5\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":2},{\\"id\\":\\"6\\",\\"name\\":\\"Mal\\",\\"kills\\":0},{\\"id\\":\\"7\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":1}],\\"totalKills\\":29},{\\"id\\":7,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":20},{\\"id\\":\\"3\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":16},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":9},{\\"id\\":\\"5\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":12},{\\"id\\":\\"6\\",\\"name\\":\\"Mal\\",\\"kills\\":-3},{\\"id\\":\\"7\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":22},{\\"id\\":\\"8\\",\\"name\\":\\"Chessus\\",\\"kills\\":0}],\\"totalKills\\":130},{\\"id\\":8,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":16},{\\"id\\":\\"3\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":20},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":12},{\\"id\\":\\"5\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":3},{\\"id\\":\\"6\\",\\"name\\":\\"Mal\\",\\"kills\\":-2},{\\"id\\":\\"7\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":10}],\\"totalKills\\":89},{\\"id\\":9,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":9},{\\"id\\":\\"3\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":0},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":12},{\\"id\\":\\"5\\",\\"name\\":\\"Chessus\\",\\"kills\\":11},{\\"id\\":\\"6\\",\\"name\\":\\"Mal\\",\\"kills\\":3},{\\"id\\":\\"7\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":10}],\\"totalKills\\":67},{\\"id\\":10,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":-1},{\\"id\\":\\"3\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":3},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":7},{\\"id\\":\\"5\\",\\"name\\":\\"Chessus\\",\\"kills\\":5},{\\"id\\":\\"6\\",\\"name\\":\\"Mal\\",\\"kills\\":1},{\\"id\\":\\"7\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":3},{\\"id\\":\\"8\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":6}],\\"totalKills\\":60},{\\"id\\":11,\\"players\\":[{\\"id\\":\\"3\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":-2},{\\"id\\":\\"2\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":5},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":0},{\\"id\\":\\"5\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":4},{\\"id\\":\\"6\\",\\"name\\":\\"Chessus\\",\\"kills\\":0},{\\"id\\":\\"7\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":-3},{\\"id\\":\\"8\\",\\"name\\":\\"Mal\\",\\"kills\\":0}],\\"totalKills\\":20},{\\"id\\":12,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":26},{\\"id\\":\\"3\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":3},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":13},{\\"id\\":\\"5\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":13},{\\"id\\":\\"6\\",\\"name\\":\\"Chessus\\",\\"kills\\":13},{\\"id\\":\\"7\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":20},{\\"id\\":\\"8\\",\\"name\\":\\"Mal\\",\\"kills\\":-6}],\\"totalKills\\":160},{\\"id\\":13,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":-1},{\\"id\\":\\"3\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":-1},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":2},{\\"id\\":\\"5\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":2},{\\"id\\":\\"6\\",\\"name\\":\\"Chessus\\",\\"kills\\":0},{\\"id\\":\\"7\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":0},{\\"id\\":\\"8\\",\\"name\\":\\"Mal\\",\\"kills\\":0}],\\"totalKills\\":6},{\\"id\\":14,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":22},{\\"id\\":\\"3\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":2},{\\"id\\":\\"4\\",\\"name\\":\\"Zeh\\",\\"kills\\":5},{\\"id\\":\\"5\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":9},{\\"id\\":\\"6\\",\\"name\\":\\"Chessus\\",\\"kills\\":7},{\\"id\\":\\"7\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":7},{\\"id\\":\\"8\\",\\"name\\":\\"Mal\\",\\"kills\\":-2}],\\"totalKills\\":122},{\\"id\\":15,\\"players\\":[{\\"id\\":\\"4\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":-3},{\\"id\\":\\"7\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":0},{\\"id\\":\\"2\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":0},{\\"id\\":\\"3\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":0},{\\"id\\":\\"5\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":0}],\\"totalKills\\":3},{\\"id\\":16,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":0},{\\"id\\":\\"3\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":0},{\\"id\\":\\"4\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":0},{\\"id\\":\\"5\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":0},{\\"id\\":\\"6\\",\\"name\\":\\"Zeh\\",\\"kills\\":0}],\\"totalKills\\":0},{\\"id\\":17,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":-2},{\\"id\\":\\"3\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":1},{\\"id\\":\\"4\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":0},{\\"id\\":\\"5\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":-3},{\\"id\\":\\"6\\",\\"name\\":\\"Zeh\\",\\"kills\\":0},{\\"id\\":\\"7\\",\\"name\\":\\"Mal\\",\\"kills\\":-1}],\\"totalKills\\":13},{\\"id\\":18,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":1},{\\"id\\":\\"3\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":0},{\\"id\\":\\"4\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":-1},{\\"id\\":\\"5\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":2},{\\"id\\":\\"6\\",\\"name\\":\\"Zeh\\",\\"kills\\":2},{\\"id\\":\\"7\\",\\"name\\":\\"Mal\\",\\"kills\\":-1}],\\"totalKills\\":7},{\\"id\\":19,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":14},{\\"id\\":\\"3\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":10},{\\"id\\":\\"4\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":14},{\\"id\\":\\"5\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":9},{\\"id\\":\\"6\\",\\"name\\":\\"Zeh\\",\\"kills\\":20},{\\"id\\":\\"7\\",\\"name\\":\\"Mal\\",\\"kills\\":2}],\\"totalKills\\":95},{\\"id\\":20,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":0},{\\"id\\":\\"3\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":1},{\\"id\\":\\"4\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":2},{\\"id\\":\\"5\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":0},{\\"id\\":\\"6\\",\\"name\\":\\"Zeh\\",\\"kills\\":0},{\\"id\\":\\"7\\",\\"name\\":\\"Mal\\",\\"kills\\":0}],\\"totalKills\\":3},{\\"id\\":21,\\"players\\":[{\\"id\\":\\"2\\",\\"name\\":\\"Isgalamido\\",\\"kills\\":17},{\\"id\\":\\"3\\",\\"name\\":\\"Oootsimo\\",\\"kills\\":22},{\\"id\\":\\"4\\",\\"name\\":\\"Dono da Bola\\",\\"kills\\":14},{\\"id\\":\\"5\\",\\"name\\":\\"Assasinu Credi\\",\\"kills\\":19},{\\"id\\":\\"6\\",\\"name\\":\\"Zeh\\",\\"kills\\":19},{\\"id\\":\\"7\\",\\"name\\":\\"Mal\\",\\"kills\\":6}],\\"totalKills\\":131}]",
  ],
]
`;
